### ISSUE TITLE
Get lease latency investigation

### ISSUE THREAD

**Issue by krzysied** [kubernetes/kubernetes#82069]
During scalability density test, it can take ~2s to get lease. This value looks wrong.

**Comment by krzysied** [comment]
Based on api call latency:
![image](https://user-images.githubusercontent.com/25119199/63842348-adc8e100-c984-11e9-83d0-87dad43955b3.png)
and etcd latency (client used by apiserver):
![image](https://user-images.githubusercontent.com/25119199/63842441-dcdf5280-c984-11e9-8b49-e92bfec7e88b.png)

It looks like the problem is with etcd.

However, in etcd logs there is not trace of call with latency > 400ms. Moreover after adding metric for grpc request handling duration:
![image](https://user-images.githubusercontent.com/25119199/63843057-16648d80-c986-11e9-82e7-3095b5dd71fd.png)
It seems there was no long request on etcd.

**Comment by krzysied** [comment]
There were tests made regarding etcd:
- Original settings:
![image](https://user-images.githubusercontent.com/25119199/63844164-fd5cdc00-c987-11e9-8588-b6f9d13caa3a.png)
- Separate nodes and leases etcds:
![image](https://user-images.githubusercontent.com/25119199/63844227-1ebdc800-c988-11e9-8064-51260843ca09.png)
- Separate nodes and leases etcd in memory:
![image](https://user-images.githubusercontent.com/25119199/63844279-3006d480-c988-11e9-9523-0a95d137501f.png)

Assuming there was a problem with etcd itself moving leases to separate etcd should fix the issue. In this case there was no visible impact.

**Comment by krzysied** [comment]
Wht's interesting the long calls are happening during saturation phase. The operations that correspond to this are secret watches.
![image](https://user-images.githubusercontent.com/25119199/63845118-ba036d00-c989-11e9-97df-30dcd395df0d.png)

**Comment by krzysied** [comment]
I ran tests with secret watch vs secret cache:
- Watch:
  - Get lease latency:
![image](https://user-images.githubusercontent.com/25119199/63850895-6a2aa300-c995-11e9-9200-36d826aeec84.png)
  - Number of active watches:
![image](https://user-images.githubusercontent.com/25119199/63850999-9cd49b80-c995-11e9-8a54-49d92f4babde.png)
  - Nomber of goroutines: 
![image](https://user-images.githubusercontent.com/25119199/63851038-b7a71000-c995-11e9-9114-328b9a995049.png)

- Cache
  - Get lease latency:
![image](https://user-images.githubusercontent.com/25119199/63851259-2dab7700-c996-11e9-9a95-359d9b0c9717.png)
  - Number of active watches:
![image](https://user-images.githubusercontent.com/25119199/63851329-664b5080-c996-11e9-989e-469b8903866d.png)
  - Number of goroutines:
![image](https://user-images.githubusercontent.com/25119199/63851627-c5a96080-c996-11e9-88e4-0ed012b0d9c7.png)

**Comment by krzysied** [comment]
> @krzysied is this the Watch vs Cache for configMapAndSecretChangeDetectionStrategy?

Yes.

**Comment by krzysied** [comment]
![image](https://user-images.githubusercontent.com/25119199/64104897-776cd680-cd75-11e9-9508-1cf6ff67d38a.png)

- Client all - e2e metric. It contains stream creation, sending msg, receiveing msg.
- Stream creation - creating new stream between apiserver and etcd.
- Client out - sending msg from apiserver to etcd
- Client in - receiveing response from etcd

**Comment by krzysied** [comment]
Comparing plots for first 30 minutes of density test:
- Watch strategy
![image](https://user-images.githubusercontent.com/25119199/64330035-88fdea80-cfd0-11e9-86d8-99a669a244f2.png)
- Cache strategy
![image](https://user-images.githubusercontent.com/25119199/64330584-82bc3e00-cfd1-11e9-97ea-a96087509794.png)
- Cache strategy that starts 3 goroutines with each get (each goroutine allocates 200B and then sleeps for 5 minutes)
![image](https://user-images.githubusercontent.com/25119199/64330792-e21a4e00-cfd1-11e9-87f6-6d9cd41295df.png)
- Cache strategy that starts 3 goroutines with each get (each goroutine allocates 100 x 200B and then sleeps for 5 minutes)
![image](https://user-images.githubusercontent.com/25119199/64330858-05dd9400-cfd2-11e9-8935-4e7da24bb85c.png)

Each plot represents phase durations of get requests. Green line is a time between sending request and receiving response header.

**Comment by wojtek-t** [comment]
xref https://github.com/kubernetes/kubernetes/issues/80209

**Comment by krzysied** [comment]
Comparing plots for first 30 minutes of density test, based on number of goroutines:
- Cache strategy - 0 goroutines
![image](https://user-images.githubusercontent.com/25119199/64343113-35999580-cfec-11e9-8b71-84a65f98ec38.png)
- Cache strategy - 3 goroutines
![image](https://user-images.githubusercontent.com/25119199/64343387-c3758080-cfec-11e9-8bf3-acacaaba2ef1.png)
- Cache strategy - 30 goroutines
![image](https://user-images.githubusercontent.com/25119199/64343550-0172a480-cfed-11e9-9709-b3739888c7c7.png)

There were `n-1` goroutines created, which were blocked on channels + one goroutine that slept for 5 minutes and then closed channels of other routines.

**Comment by wojtek-t** [comment]
Good findings. So the next step is to probably check whether with smaller number of goroutines (like 3), but allocating visibly more memory (a bunch of smaller and larger chunks) how it looks.

**Comment by krzysied** [comment]
- Watch cache
![image](https://user-images.githubusercontent.com/25119199/64420588-ea40bf00-d09f-11e9-90c9-ee4d2f223f07.png)
- Watch cache with reduced number of goroutines - 400k instead of 600k
![image](https://user-images.githubusercontent.com/25119199/64420663-0e9c9b80-d0a0-11e9-9396-2eb07e4e7faf.png)

**Comment by krzysied** [comment]
Watch strategy with fakeWatch (for test pod's secrets) - 450 goroutines
![image](https://user-images.githubusercontent.com/25119199/64440270-f80c3980-d0cb-11e9-8c93-2b3ac04b9976.png)

**Comment by krzysied** [comment]
Watch strategy with fake ServerWatch - 300 goroutines
![image](https://user-images.githubusercontent.com/25119199/64520710-fe322e00-d2f6-11e9-8ac0-dcc8a73e8229.png)

This time the only goroutines created are request handlers. There was only header response sent to establish connection.

**Comment by krzysied** [comment]
Watch strategy with fake watch on client side:
![image](https://user-images.githubusercontent.com/25119199/64553149-26db1780-d339-11e9-85b3-6f2216812ee1.png)
Watch strategy is listing secrets.

**Comment by wojtek-t** [comment]
I don't understand the last one - what does it mean? It means "no requests are send"? Then what was the purpose of this experiment?

**Comment by krzysied** [comment]
We know that changing strategy has impact on get lease latency. IIUC this means that we are listing secrets (each secret once as initialization) and than we are creating watch. I wanted to check if listing secrets (instead of using get) has any impact on lease latency.

**Comment by krzysied** [comment]
Accessing custom made endpoint that call sleep for 5 min. Connection is being kept alive during this time.
![image](https://user-images.githubusercontent.com/25119199/64637157-856cc800-d403-11e9-8ab9-af7769564441.png)

**Comment by krzysied** [comment]
- First run of density test
![image](https://user-images.githubusercontent.com/25119199/64697289-cc5ccb00-d4a0-11e9-9628-3772a5bad8b5.png)
- Second run of density test (same apiserver)
![image](https://user-images.githubusercontent.com/25119199/64699512-35464200-d4a5-11e9-8e0b-072b397facbf.png)

**Comment by wojtek-t** [comment]
Given that I requested this experiment, let me explain the above (because this is expected - I just wanted to see confirmation).
Based on what I've learned from Go folks: when we create a goroutine it is allocating some memory for itself and some memory for its stack. When goroutine finishes, the former stays forever (it's not GC-ed) [though obviously can be reused by future goroutines].

So what we observe here is that we are relatively quickly increasing number of goroutines which (together with other memory allocations we're doing) is almost for sure causing contention on memory allocations. In the second run, we only reuse already allocated memory for goroutines, thus the contention is much smaller (and as a result results are much better).

**Comment by krzysied** [comment]
Calling sleep method on standard server, without k8s default filer chain.
![image](https://user-images.githubusercontent.com/25119199/64952476-bb390300-d880-11e9-8014-51f456b04c99.png)

I'll be running test with separate http2 server. However, IMHO it seems that problem is on creating new stream handler level and not on handling requests by k8s level.

**Comment by krzysied** [comment]
Calling sleep on separate server using http2. There was only one connection made per node. All sleep calls were send using streams.
![image](https://user-images.githubusercontent.com/25119199/65061397-1ef32700-d97a-11e9-9c07-d2fb0b71d7b0.png)
 IMHO it looks the same as calling standard k8s server.

**Comment by wojtek-t** [comment]
/remove-lifecycle stale

I think we already understand everything is happening here - I will try to summarize it tomorrow.

**Comment by wojtek-t** [comment]
To quickly summarize it:
- the high latencies were caused by lock contention for memory allocations at golang level
- memory allocations were originating from creating new watches

The problem was happening in density test, when we are quickly increasing the number of active watches in the system (which means quickly growing memory usage at apiserver). That is resulting in lock contention at the golang level (when allocating new page from operating system).
The watches are coming from kubelets watching corresponding  secrets/configmaps,

The outcomes from this issue are:
- we are already working with golang folks to understand how much can be done on golang side to improve the situation - though that is a longer-term effort
- I've opened a proposal (that is almost approved already) to allow for marking secrets/config maps as immutable (on opt-in basis), to mitigate the impact

We're not going to do anything else as part of this bug for now, so closing it.
